### PR TITLE
Consider surrogate pair for moving(h/l) and yank

### PR DESCRIPTION
**Pull request by smallkirby**
**What this PR does / why we need it**:

This PR is related to #6046 .

This fix adds a check if the character the cursor is moving to  is a part of surrogate pairs. If the character is a part of surrogate pairs, it additionally moves a cursor.

Before this PR, yank and moving for surrogate pairs look like below (the first character is `𩸽`, and the second is `🐕`):

https://user-images.githubusercontent.com/47111091/158623636-d5a42600-f001-4f7b-9ffd-023f1beb8775.mp4

After this PR, it looks like below:

https://user-images.githubusercontent.com/47111091/158623701-ad5c1bd8-9e8d-4087-9f65-a6ae96ec78ab.mp4

**Which issue(s) this PR fixes**

Related to #6046, but I don't know this fixes it or not.

**Special notes for your reviewer**:

- I'm sorry for dirty commits which just fixes typo. If it is bad, I will rebase them, or open new PR.

### Environment of PoC movie above
- VSCode version: 1.64.2
- OS: Linux x64 5.16.9-custom (Ubuntu Jammy Jellyfish (devel branch))

### CODE REVIEW

**Comment by J-Fields** [comment]
Thanks for the PR!

> I'm sorry for dirty commits which just fixes typo. If it is bad, I will rebase them, or open new PR.

No trouble at all; except for very large PRs, I tend to ignore individual commits and just squash them at the end.

Would you mind adding a few test cases, please?

**Comment by smallkirby** [comment]
Thanks for your quick reply. I added a test suite which tests moving and yank of surrogate-pairs.

**Comment by smallkirby** [comment]
Let me summarize a little bit more in detail.

- Surrogate pairs are pairs of two bytes, which consist of a single Unicode character, such as Emoji and Kanji. Surrogate pairs consist of `high surrogate` and `low surrogate`. In VSCodeVim, these pairs are just treated as two separate characetrs when moving cursor or yanking characters.
  - Therefore as #6894 and #2722, you need to type `ll` to move left across surrogate pairs.
  - Yanking surrogate pairs by `vy` just copies a single byte of the pair. It is pasted to as invalid character.

- In this PR, I managed that cursor position must always be at `high surrogate`, not `low surrogate` for navigation.
  - If the cursor is located at `low surrogate` byte of surrogate pairs, move left or right additionaly depending on navigation direction.

- Same for yank. If the start position of yank is located at `low surrogate`, move start position to left. If the end position of yank is located at `high surrogate`, move end position to right.
  - In short, range of selection should not laid across `high surrogate` and `low surrogate`.

- The largest impact for users would be that surrogate pairs are treated as single character for navigation. It might need discussion though vanilla vim (I mean vim-tiny or something) treats surrogate pairs in the same way as far as I know.

- As for #6046, I don't know whether this PR fixes it or not, because I can't reproduce it. I can copy `𩸽` outside VSCodeVim, and paste it to VSCode without any modification.

**Review by J-Fields — APPROVED** [review]
LGTM. Thank you!
